### DISCUSSION TITLE
help with Libigl's variadic booleans

### DISCUSSION THREAD

**Discussion by alecjacobson** [libigl/libigl#2033] [Q&A]
From email:

> I hope I don't find you too busy. I am writing to you because I am trying to implement a variadic boolean in libigl, with no success so far. 
> 
> I see in igl/copyleft/cgal/mesh_boolean.h there is a function call for variadics but I cannot really understand how it works. Can you please help me out on this?
> 
> Suppose I have meshes A,B,C and I'd like to obtain A minus (B,C). What should my keep() and wind_num_func() functions be like?
> 
> Thanks in advance for your help!
>

**Comment by alecjacobson** [comment]
For solid geometry operations the `keep()` function can be generated via `KeepInside()` in `#include <igl/copyleft/cgal/BinaryWindingNumberOperations.h>`. This simply indicates that the resulting mesh should retain the faces corresponding to the boundary of the solid interior of the resulting operation.

The `wind_num_func()` should be a function that takes in a vector of winding number values where the ith element corresponds to the winding number of the ith input shape. This vector will be constant for each "cell" of the arrangement (intersection resolution) of the input shapes. For solid operations, this function should return a boolean whether the cell is deemed _inside_ based on its winding number vector. 

Here's an complete example program that reads in n meshes and  subtracts from the first object the union of all other objects `X₀ \ (X₁ ∪ X₂ ∪ …)`.

If you run this like `example {cube,xcylinder,ycylinder,zcylinder}.obj` then you should see the inputs in wireframe and the output surface visualized:

<img width="574" alt="Screen Shot 2022-05-02 at 11 38 34 AM" src="https://user-images.githubusercontent.com/2241689/166263082-bb1e22e4-9cde-4718-9aec-d85400fa554d.png">


```cpp
#include <igl/read_triangle_mesh.h>
#include <igl/copyleft/cgal/mesh_boolean.h>
#include <igl/opengl/glfw/Viewer.h>

#include <igl/copyleft/cgal/BinaryWindingNumberOperations.h>

int main(int argc, char * argv[])
{
  // Read in inputs and add to viewer in wireframe
  igl::opengl::glfw::Viewer vr;
  std::vector<Eigen::MatrixXd> V;
  std::vector<Eigen::MatrixXi> F;
  for(int i = 0;i<argc-1;i++)
  {
    printf("%d: %s\n",i,argv[i+1]);
    V.emplace_back();
    F.emplace_back();
    igl::read_triangle_mesh(argv[i+1],V[i],F[i]);
    
    vr.data().set_mesh(V[i],F[i]);
    vr.data().show_lines = true;
    vr.data().show_faces = false;
    vr.append_mesh();
  }

  // Winding number function for subtracting the union of all other objects from
  // the first object.
  //
  // X₀ \ (X₁ ∪ X₂ ∪ …)
  std::function<int(const Eigen::Matrix<int,1,Eigen::Dynamic>) >
    wind_num_op = [](const Eigen::Matrix<int,1,Eigen::Dynamic> & win_nums)->int
    {
      assert(win_nums.size()>1);
      // Union of objects 1 through n-1
      bool union_rest = false;
      for(int i = 1;i<win_nums.size();i++)
      {
        union_rest = union_rest || win_nums(i) > 0;
        if(union_rest) break;
      }
      // Must be in object 0 and not in union of objects 1 through n-1
      return win_nums(0) > 0 && !union_rest;
    };
  // Conduct variadic boolean operation into VO,FO with J indicating origins
  Eigen::MatrixXd VO;
  Eigen::MatrixXi FO;
  Eigen::VectorXi J;
  igl::copyleft::cgal::mesh_boolean(
    V,F,wind_num_op, igl::copyleft::cgal::KeepInside(),VO,FO,J);

  // Visualize output mesh
  vr.data().set_mesh(VO,FO);
  vr.data().set_data(J.cast<double>());
  vr.data().show_lines = false;
  vr.launch();
}
```



[meshes.zip](https://github.com/libigl/libigl/files/8603976/meshes.zip)

**Reply by mlivesu** [reply]
thanks Alec!
